src/server.c

```diff
@@ -4061,6 +4045,52 @@ uint64_t getCommandFlags(client *c) {
     return cmd_flags;
 }
 
+void preprocessCommand(client *c, pendingCommand *pcmd) {
+    pcmd->slot = INVALID_CLUSTER_SLOT;
+    if (pcmd->argc == 0)
+        return;
+
+    /* Check if we can reuse the last command instead of looking it up.
+     * The last command is either the penultimate pending command (if it exists), or c->lastcmd. */
+    struct redisCommand *last_cmd = c->pending_cmds.tail->prev ? c->pending_cmds.head->cmd : c->lastcmd;
+
+    if (isCommandReusable(last_cmd, pcmd->argv[0]))
+        pcmd->cmd = last_cmd;
+    else
+        pcmd->cmd = lookupCommand(pcmd->argv, pcmd->argc);
+
+    if (!pcmd->cmd) {
+        pcmd->flags = CLIENT_READ_COMMAND_NOT_FOUND;
+        return;
+    }
+
+    if ((pcmd->cmd->arity > 0 && pcmd->cmd->arity != pcmd->argc) ||
+        (pcmd->argc < -pcmd->cmd->arity))
+    {
+        pcmd->flags = CLIENT_READ_BAD_ARITY;
+        return;
+    }
+
+    if (server.cluster_enabled) {
+        getKeysResult result = (getKeysResult)GETKEYS_RESULT_INIT;
+        int numkeys = getKeysFromCommand(pcmd->cmd, pcmd->argv, pcmd->argc, &result); 
```

Comment:
since this is conditional here, it means ACL can't re-use it and if both cluster mode and ACL are in used, this part is done twice.